src/main/java/com/icthh/xm/uaa/security/oauth2/idp/config/IdpConfigRepository.java

```diff
@@ -0,0 +1,234 @@
+
+/**
+ * This class reads and process IDP clients public configuration for each tenant.
+ * Tenant IDP clients created for each successfully loaded and valid config. If config not present or invalid it skipped.
+ */
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class IdpConfigRepository implements RefreshableConfiguration {
+
+    private static final String KEY_TENANT = "tenant";
+
+    private final ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
+    private final AntPathMatcher matcher = new AntPathMatcher();
+
+    /**
+     * In memory storage for storing information tenant IDP clients public.
+     * idpClientConfigs structure Map< tenantKey, Map< clientKey, IdpPublicClientConfig>> where:
+     * <ul>
+     *   <li/> tenantKey - tenant id
+     *   <li/> clientKey - client id
+     *   <li/> IdpPublicClientConfig - tenant specific idp public client Config
+     * </ul>
+     */
+    private final Map<String, Map<String, IdpPublicClientConfig>> idpClientConfigs = new ConcurrentHashMap<>();
+
+    /**
+     * In memory storage.
+     * Stores information about tenant IDP clients public configuration that currently in process.
+     * <p/>
+     * We need to store this information in memory to avoid corruption previously registered in-memory tenant clients config
+     * tmpIdpClientPublicConfigs structure Map< tenantKey, Map< clientKey, IdpPublicClientConfig>> where:
+     * <ul>
+     *   <li/> tenantKey - tenant id
+     *   <li/> clientKey - client id
+     *   <li/> IdpPublicClientConfig - tenant specific idp public client Config
+     * </ul>
+     */
+    private final Map<String, Map<String, IdpPublicClientConfig>> tmpValidIdpClientPublicConfigs = new ConcurrentHashMap<>();
+
+    /**
+     * In memory storage.
+     * Stores information about tenant IDP jwt claims verifiers.
+     * <p/>
+     * We need to store this information in memory to avoid initialization on each token claims verification
+     * and store actual verifiers.
+     * <p/>
+     * jwtClaimsSetVerifiersHolder structure Map< tenantKey, Map< clientKey, List< JwtClaimsSetVerifier >>> where:
+     * <ul>
+     *   <li/> tenantKey - tenant id
+     *   <li/> clientKey - client id
+     *   <li/> List< JwtClaimsSetVerifier> - list of default token claim verifiers
+     * </ul>
+     */
+    private final Map<String, Map<String, List<JwtClaimsSetVerifier>>> jwtClaimsSetVerifiersHolder = new ConcurrentHashMap<>();
+
+    private final TenantContextHolder tenantContextHolder;
+
+    @Override
+    public boolean isListeningConfiguration(String updatedKey) {
+        return matcher.match(IDP_PUBLIC_SETTINGS_CONFIG_PATH_PATTERN, updatedKey);
+    }
+
+    @Override
+    public void onInit(String configKey, String configValue) {
+        updateIdpConfigs(configKey, configValue);
+    }
+
+    @Override
+    public void onRefresh(String updatedKey, String config) {
```

Comment:
handle case when config was removed

Reply:
handled in 
`if (isRawClientsConfigurationEmpty && isValidClientsConfigurationEmpty) {
                log.warn("For tenant [{}] provided IDP public client configs not present." +
                    "Removing client configs and claim verifiers from storage", tenantKey);
                idpClientConfigs.remove(tenantKey);
                jwtClaimsSetVerifiersHolder.remove(tenantKey);
                return;
            }`